subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1065 @@
+		pipe->bytes_used, pipe->read_index, pipe->write_index);
+
+	if (pipe->read_index == pipe->write_index) {
+		return pipe->bytes_used;
+	} else if (pipe->read_index < pipe->write_index) {
+		return pipe->write_index - pipe->read_index;
+	}
+
+	return pipe->size - (pipe->read_index - pipe->write_index);
+}
+
+/**
+ * Determine bytes available to write
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * written to a given @ref spair without blocking
+ *
+ * The calling thread must have exclusive access over the @ref spair.
+ */
+static inline size_t spair_write_avail(const struct spair *spair)
+{
+	const struct spair *remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	if (!remote) {
+		return 0;
+	}
+
+	return k_pipe_write_avail(&remote->recv_q);
+}
+
+/**
+ * Determine bytes available to read
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * written to a given @ref spair without blocking
+ *
+ * The calling thread must have exclusive access over the @ref spair.
+ */
+static inline size_t spair_read_avail(const struct spair *spair)
+{
+	return k_pipe_read_avail(&spair->recv_q);
+}
+
+/**
+ * Swap two 32-bit integers
+ */
+static inline void swap32(u32_t *a, u32_t *b)
+{
+	u32_t c;
+
+	c = *b;
+	*b = *a;
+	*a = c;
+}
+
+/**
+ * Delete a @ref spair (1/2 of a socketpair)
+ *
+ * Note, if the @ref spair.remote file descriptor field is not equal to -1,
+ * then this function will attempt to free it.
+ *
+ * That is done so that this function can be called from @ref spair_delete
+ * or @ref z_impl_zsock_socketpair in case there is an initialization
+ * error.
+ */
+static void spair_delete(struct spair *spair)
+{
+	if (!spair) {
+		return;
+	}
+
+	if (-1 != spair->remote) {
+		z_free_fd(spair->remote);
+		spair->remote = -1;
+	}
+
+	memset(spair, 0, sizeof(*spair));
+	k_free(spair);
+	spair = NULL;
+}
+
+/**
+ * Create a @ref spair (1/2 of a socketpair)
+ *
+ * The idea is to call this twice, but store the "local" side in the
+ * @ref spair.remote field initially.
+ *
+ * If both allocations are successful, then swap the @ref spair.remote
+ * fields in the two @ref spair instances.
+ */
+static struct spair *spair_new(void)
+{
+	struct spair *spair;
+
+	spair = k_malloc(sizeof(*spair));
+	if (!spair) {
+		errno = ENOMEM;
+		goto out;
+	}
+	memset(spair, 0, sizeof(*spair));
+
+	/* initialize any non-zero default values */
+	spair->remote = -1;
+	spair->flags = SPAIR_FLAGS_DEFAULT;
+
+	k_sem_init(&spair->sem, 0, 1);
+	k_pipe_init(&spair->recv_q, spair->buf, sizeof(spair->buf));
+	k_poll_signal_init(&spair->write_signal);
+	k_poll_signal_init(&spair->read_signal);
+
+	spair->remote = z_reserve_fd();
+	if (-1 == spair->remote) {
+		errno = ENFILE;
+		goto cleanup;
+	}
+
+	z_finalize_fd(spair->remote, spair,
+		      (const struct fd_op_vtable *)&spair_fd_op_vtable);
+
+	LOG_DBG("%p fd: %d sem: %p vtable: %p recv_q: %p buf: %p "
```

Comment:
In networking code, it would be better to use `NET_DBG()`. The network variant adds thread id to debug print which is quite useful in debugging.